File: reports/weight.go
package reports

import (
	"strings"
	"time"
)

// ClassifyWeight assigns a heuristic cost tier (LOW, MEDIUM, HEAVY) to a
// report definition so callers can throttle or warn before executing expensive
// queries.
func ClassifyWeight(def ReportDefinition) string {
	// Time-series + secondary group-by: time pivot (most expensive case first)
	if def.SecondaryGroupBy != "" && def.Mode == "time_series" {
		eitherIsArtifact := isArtifactDimension(def.GroupBy) || isArtifactDimension(def.SecondaryGroupBy)
		if eitherIsArtifact {
			if hasDateFilter(def.Filters) {
				return "MEDIUM"
			}
			return "HEAVY"
		}
		if !hasDateFilter(def.Filters) {
			return "HEAVY"
		}
		if dateFilterWindowDays(def.Filters) > 90 {
			return "HEAVY"
		}
		return "MEDIUM"
	}

	// Aggregate 2D pivot
	if def.SecondaryGroupBy != "" {
		eitherIsArtifact := isArtifactDimension(def.GroupBy) || isArtifactDimension(def.SecondaryGroupBy)
		if eitherIsArtifact {
			if hasDateFilter(def.Filters) {
				return "MEDIUM"
			}
			return "HEAVY"
		}
		return "LOW"
	}

	// 1D time series
	if def.Mode == "time_series" {
		if def.TimeBucket == "custom" {
			days := customBucketDays(def.CustomBucketN, def.CustomBucketUnit)
			if days > 90 {
				return "HEAVY"
			}
			return "MEDIUM"
		}
		if !hasDateFilter(def.Filters) {
			return "HEAVY"
		}
		days := dateFilterWindowDays(def.Filters)
		if days > 90 {
			return "HEAVY"
		}
		return "MEDIUM"
	}

	if def.Mode == "aggregate" && hasArtifactScopeFilter(def.Filters) {
		return "MEDIUM"
	}
	return "LOW"
}

func customBucketDays(n int, unit string) int {
	switch unit {
	case "day":
		return n
	case "week":
		return n * 7
	case "month":
		return n * 30
	}
	return n
}

func hasDateFilter(f ReportFilters) bool {
	for _, c := range f.And {
		if c.TopLevel == "launchDT" || c.TopLevel == "returnDT" {
			return true
		}
	}
	for _, group := range f.Or {
		for _, c := range group {
			if c.TopLevel == "launchDT" || c.TopLevel == "returnDT" {
				return true
			}
		}
	}
	return false
}

// dateFilterWindowDays returns a rough estimate of the filter window in days.
// Returns a large sentinel (9999) if the window cannot be determined.
func dateFilterWindowDays(f ReportFilters) int {
	minDays := 9999
	all := make([]FilterCondition, 0, len(f.And))
	all = append(all, f.And...)
	for _, group := range f.Or {
		all = append(all, group...)
	}
	now := time.Now()
	for _, c := range all {
		if c.TopLevel != "launchDT" && c.TopLevel != "returnDT" {
			continue
		}
		if c.Op != ">" && c.Op != ">=" {
			continue
		}
		t, err := time.Parse("2006-01-02", c.Val)
		if err != nil {
			continue
		}
		days := int(now.Sub(t).Hours() / 24)
		if days < minDays {
			minDays = days
		}
	}
	return minDays
}

func hasArtifactScopeFilter(f ReportFilters) bool {
	isArtifactField := func(s string) bool { return strings.HasPrefix(s, "artifact_") }
	for _, c := range f.And {
		if isArtifactField(c.TopLevel) {
			return true
		}
	}
	for _, group := range f.Or {
		for _, c := range group {
			if isArtifactField(c.TopLevel) {
				return true
			}
		}
	}
	return false
}

func isArtifactDimension(groupBy string) bool {
	switch groupBy {
	case "artifact_name", "rarity", "tier", "spec_type":
		return true
	}
	return false
}

File: src/EggLedger.Domain/Reports/Weight.cs
using System.Globalization;

namespace EggLedger.Domain.Reports;

/// <summary>
/// Heuristic query-cost classification for report definitions. Port of Go
/// reports/weight.go. Pure aside from the wall-clock used by date-window sizing.
/// </summary>
public static class Weight
{
    /// <summary>
    /// Assigns a cost tier ("LOW", "MEDIUM", "HEAVY") to a report definition.
    /// Port of Go ClassifyWeight.
    /// </summary>
    public static string ClassifyWeight(ReportDefinition def)
    {
        // Time-series + secondary group-by: time pivot (most expensive case first).
        if (def.SecondaryGroupBy != "" && def.Mode == "time_series")
        {
            var eitherIsArtifact = IsArtifactDimension(def.GroupBy) || IsArtifactDimension(def.SecondaryGroupBy);
            if (eitherIsArtifact)
            {
                if (HasDateFilter(def.Filters))
                {
                    return "MEDIUM";
                }
                return "HEAVY";
            }
            if (!HasDateFilter(def.Filters))
            {
                return "HEAVY";
            }
            if (DateFilterWindowDays(def.Filters) > 90)
            {
                return "HEAVY";
            }
            return "MEDIUM";
        }

        // Aggregate 2D pivot.
        if (def.SecondaryGroupBy != "")
        {
            var eitherIsArtifact = IsArtifactDimension(def.GroupBy) || IsArtifactDimension(def.SecondaryGroupBy);
            if (eitherIsArtifact)
            {
                if (HasDateFilter(def.Filters))
                {
                    return "MEDIUM";
                }
                return "HEAVY";
            }
            return "LOW";
        }

        // 1D time series.
        if (def.Mode == "time_series")
        {
            if (def.TimeBucket == "custom")
            {
                var days = CustomBucketDays(def.CustomBucketN, def.CustomBucketUnit);
                if (days > 90)
                {
                    return "HEAVY";
                }
                return "MEDIUM";
            }
            if (!HasDateFilter(def.Filters))
            {
                return "HEAVY";
            }
            var d = DateFilterWindowDays(def.Filters);
            if (d > 90)
            {
                return "HEAVY";
            }
            return "MEDIUM";
        }

        if (def.Mode == "aggregate" && HasArtifactScopeFilter(def.Filters))
        {
            return "MEDIUM";
        }
        return "LOW";
    }

    private static int CustomBucketDays(int n, string unit) => unit switch
    {
        "day" => n,
        "week" => n * 7,
        "month" => n * 30,
        _ => n,
    };

    private static bool HasDateFilter(ReportFilters f)
    {
        foreach (var c in f.And)
        {
            if (c.TopLevel == "launchDT" || c.TopLevel == "returnDT")
            {
                return true;
            }
        }
        foreach (var group in f.Or)
        {
            foreach (var c in group)
            {
                if (c.TopLevel == "launchDT" || c.TopLevel == "returnDT")
                {
                    return true;
                }
            }
        }
        return false;
    }

    /// <summary>
    /// Rough estimate of the filter window in days. Returns 9999 if it cannot be
    /// determined. Port of Go dateFilterWindowDays.
    /// </summary>
    private static int DateFilterWindowDays(ReportFilters f)
    {
        var minDays = 9999;
        var all = new List<FilterCondition>(f.And);
        foreach (var group in f.Or)
        {
            all.AddRange(group);
        }
        var now = DateTime.Now;
        foreach (var c in all)
        {
            if (c.TopLevel != "launchDT" && c.TopLevel != "returnDT")
            {
                continue;
            }
            if (c.Op != ">" && c.Op != ">=")
            {
                continue;
            }
            if (!DateTime.TryParseExact(c.Val, "yyyy-MM-dd", CultureInfo.InvariantCulture,
                    DateTimeStyles.None, out var t))
            {
                continue;
            }
            var days = (int)((now - t).TotalHours / 24);
            if (days < minDays)
            {
                minDays = days;
            }
        }
        return minDays;
    }

    private static bool HasArtifactScopeFilter(ReportFilters f)
    {
        static bool IsArtifactField(string s) => s.StartsWith("artifact_", StringComparison.Ordinal);
        foreach (var c in f.And)
        {
            if (IsArtifactField(c.TopLevel))
            {
                return true;
            }
        }
        foreach (var group in f.Or)
        {
            foreach (var c in group)
            {
                if (IsArtifactField(c.TopLevel))
                {
                    return true;
                }
            }
        }
        return false;
    }

    private static bool IsArtifactDimension(string groupBy) => groupBy switch
    {
        "artifact_name" or "rarity" or "tier" or "spec_type" => true,
        _ => false,
    };
}
